File: omrchecker-js/packages/core/src/processors/threshold/LocalThreshold.ts
/**
 * Local threshold strategy for bubble detection.
 *
 * TypeScript port of src/processors/threshold/local_threshold.py
 * Maintains 1:1 correspondence with Python implementation.
 */

import { ThresholdConfig, ThresholdResult } from './GlobalThreshold';
import { Logger } from '../../utils/logger';

const logger = new Logger('LocalThreshold');

/**
 * Local threshold strategy - calculates threshold per question/group.
 *
 * Strategy using field-level statistics.
 * Based on the existing get_local_threshold logic.
 * Calculates threshold for individual field, with fallback to global.
 *
 * This method is useful when lighting conditions vary across the image.
 */
export class LocalThreshold {
  private globalFallback: number | null;

  /**
   * Initialize with optional global fallback threshold.
   *
   * @param globalFallback - Global threshold to use when local confidence is low
   */
  constructor(globalFallback: number | null = null) {
    this.globalFallback = globalFallback;
  }

  /**
   * Calculate local threshold with global fallback.
   *
   * @param bubbleMeanValues - List of bubble mean intensity values
   * @param config - Threshold configuration
   * @returns ThresholdResult with threshold and confidence information
   */
  calculateThreshold(
    bubbleMeanValues: number[],
    config: ThresholdConfig
  ): ThresholdResult {
    const fallbackThreshold =
      this.globalFallback ?? config.defaultThreshold;

    // Base case: empty or single bubble
    if (bubbleMeanValues.length < 2) {
      return {
        thresholdValue: fallbackThreshold,
        confidence: 0.0,
        maxJump: 0.0,
        methodUsed: 'local_single_bubble_fallback',
        fallbackUsed: true,
        metadata: {},
      };
    }

    const sortedValues = [...bubbleMeanValues].sort((a, b) => a - b);

    // Special case: exactly 2 bubbles
    if (sortedValues.length === 2) {
      const gap = sortedValues[1] - sortedValues[0];
      const minGapTwoBubbles = config.minGapTwoBubbles ?? 20.0;

      if (gap < minGapTwoBubbles) {
        return {
          thresholdValue: fallbackThreshold,
          confidence: 0.3,
          maxJump: gap,
          methodUsed: 'local_two_bubbles_small_gap_fallback',
          fallbackUsed: true,
          metadata: {},
        };
      }

      // Calculate mean of the two values
      const mean = (sortedValues[0] + sortedValues[1]) / 2;

      return {
        thresholdValue: mean,
        confidence: 0.7,
        maxJump: gap,
        methodUsed: 'local_two_bubbles_mean',
        fallbackUsed: false,
        metadata: {},
      };
    }

    // 3+ bubbles: find largest jump
    let maxJump = 0.0;
    let threshold = fallbackThreshold;

    for (let i = 1; i < sortedValues.length - 1; i++) {
      const jump = sortedValues[i + 1] - sortedValues[i - 1];
      if (jump > maxJump) {
        maxJump = jump;
        threshold = sortedValues[i - 1] + jump / 2;
      }
    }

    // Check if jump is confident
    const minJumpSurplusForGlobalFallback =
      config.minJumpSurplusForGlobalFallback ?? 10.0;
    const confidentJump = config.minJump + minJumpSurplusForGlobalFallback;

    if (maxJump < confidentJump) {
      // Low confidence - use global fallback
      return {
        thresholdValue: fallbackThreshold,
        confidence: 0.4,
        maxJump: maxJump,
        methodUsed: 'local_low_confidence_global_fallback',
        fallbackUsed: true,
        metadata: { localThreshold: threshold },
      };
    }

    // High confidence
    const confidence = Math.min(1.0, maxJump / (confidentJump * 2));

    logger.debug(
      `LocalThreshold: maxJump=${maxJump.toFixed(2)}, threshold=${threshold.toFixed(2)}, confidence=${confidence.toFixed(2)}`
    );

    return {
      thresholdValue: threshold,
      confidence: confidence,
      maxJump: maxJump,
      methodUsed: 'local_max_jump',
      fallbackUsed: false,
      metadata: { numBubbles: bubbleMeanValues.length },
    };
  }
}

File: src/processors/detection/threshold/local_threshold.py
"""Local threshold strategy for bubble detection.

Strategy using field-level statistics.
Based on the existing get_local_threshold logic.
Calculates threshold for individual field, with fallback to global.
"""

import numpy as np

from src.processors.detection.threshold.threshold_result import (
    ThresholdConfig,
    ThresholdResult,
)
from src.processors.detection.threshold.threshold_strategy import ThresholdStrategy


class LocalThresholdStrategy(ThresholdStrategy):
    """Strategy using field-level statistics.

    Based on the existing get_local_threshold logic.
    Calculates threshold for individual field, with fallback to global.
    """

    def __init__(self, global_fallback: float | None = None) -> None:
        """Initialize with optional global fallback threshold.

        Args:
            global_fallback: Global threshold to use when local confidence is low
        """
        self.global_fallback = global_fallback

    def calculate_threshold(
        self, bubble_mean_values: list[float], config: ThresholdConfig
    ) -> ThresholdResult:
        """Calculate local threshold with global fallback."""
        fallback_threshold = self.global_fallback or config.default_threshold

        # Base case: empty or single bubble
        if len(bubble_mean_values) < 2:
            return ThresholdResult(
                threshold_value=fallback_threshold,
                confidence=0.0,
                max_jump=0.0,
                method_used="local_single_bubble_fallback",
                fallback_used=True,
            )

        sorted_values = sorted(bubble_mean_values)

        # Special case: exactly 2 bubbles
        if len(sorted_values) == 2:
            gap = sorted_values[1] - sorted_values[0]
            if gap < config.min_gap_two_bubbles:
                return ThresholdResult(
                    threshold_value=fallback_threshold,
                    confidence=0.3,
                    max_jump=gap,
                    method_used="local_two_bubbles_small_gap_fallback",
                    fallback_used=True,
                )
            return ThresholdResult(
                threshold_value=float(np.mean(sorted_values)),
                confidence=0.7,
                max_jump=gap,
                method_used="local_two_bubbles_mean",
                fallback_used=False,
            )

        # 3+ bubbles: find largest jump
        max_jump = 0.0
        threshold = fallback_threshold

        for i in range(1, len(sorted_values) - 1):
            jump = sorted_values[i + 1] - sorted_values[i - 1]
            if jump > max_jump:
                max_jump = jump
                threshold = sorted_values[i - 1] + jump / 2

        # Check if jump is confident
        confident_jump = config.min_jump + config.min_jump_surplus_for_global_fallback

        if max_jump < confident_jump:
            # Low confidence - use global fallback
            return ThresholdResult(
                threshold_value=fallback_threshold,
                confidence=0.4,
                max_jump=max_jump,
                method_used="local_low_confidence_global_fallback",
                fallback_used=True,
                metadata={"local_threshold": threshold},
            )

        # High confidence
        confidence = min(1.0, max_jump / (confident_jump * 2))

        return ThresholdResult(
            threshold_value=threshold,
            confidence=confidence,
            max_jump=max_jump,
            method_used="local_max_jump",
            fallback_used=False,
            metadata={"num_bubbles": len(bubble_mean_values)},
        )
